src/main/flight/autopilot.c

```diff
@@ -91,14 +158,223 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+        float accelerationEW = (velocityEW - posHold.previousVelocityEW) * posHold.gpsDataFreqHz;
+        posHold.previousVelocityEW = velocityEW;
+
+        // scale and filter - filter cutoffs vary during the startup phase
+        float nsD = velocityNS * positionPidCoeffs.Kd;
+        pt1FilterUpdateCutoff(&velocityNSLpf, posHold.pt1Gain);
+        nsD = pt1FilterApply(&velocityNSLpf, nsD);
+
+        float ewD = velocityEW * positionPidCoeffs.Kd;
+        pt1FilterUpdateCutoff(&velocityEWLpf, posHold.pt1Gain);
+        ewD = pt1FilterApply(&velocityEWLpf, ewD);
+
+        float nsA = accelerationNS * positionPidCoeffs.Kf;
+        pt2FilterUpdateCutoff(&accelerationNSLpf, posHold.pt1Gain); // using PT1 gain for stronger cutoff
+        nsA = pt2FilterApply(&accelerationNSLpf, nsA);
+
+        float ewA = accelerationEW * positionPidCoeffs.Kf;
+        pt2FilterUpdateCutoff(&accelerationEWLpf, posHold.pt1Gain); // using PT1 gain for stronger cutoff
+        ewA = pt2FilterApply(&accelerationEWLpf, ewA);
+
+        // limit sum of D and A because otherwise can be too aggressive when starting at speed
+        float nsDA = nsD + nsA;
+        float ewDA = ewD + ewA;
+        const float maxDAAngle = 35.0f; // limit in degrees; arbitrary.  20 is a bit too low, allows a lot of overshoot
+        // an angle of more than 35 degrees is achieved as P and I grow
+        // ** todo = should this be half of the user-configurable angle_limit?  Or fixed?
+        nsDA = constrainf(nsDA, -maxDAAngle, maxDAAngle);
+        ewDA = constrainf(ewDA, -maxDAAngle, maxDAAngle);
+
+        // ** PID Sum **
+
+        float nsPidSum = nsP + nsI + nsDA;
+        float ewPidSum = ewP + ewI + ewDA;
+
+        // terminate initial startup behaviour as pidSum crosses zero
+        // this is best for smoothness of the transition from the stop to the new target location
+        if (posHold.isStartingNS && posHold.previousPidSumNS * nsPidSum < 0.0f) { // pidsum ns has reversed sign
+            resetPositionControlParamsNS();
+            currentTargetLocation.lat = gpsSol.llh.lat;
+            nsDistance = 0.0f;
+            posHold.isStartingNS = false;
+        }
+        if (posHold.isStartingEW && posHold.previousPidSumEW * ewPidSum < 0.0f) { // pidsum ns has reversed sign
+            resetPositionControlParamsEW();
+            ewDistance = 0.0f;
+            currentTargetLocation.lon = gpsSol.llh.lon;
+            posHold.isStartingEW = false;
+        }
+        posHold.previousPidSumNS = nsPidSum;
+        posHold.previousPidSumEW = ewPidSum;
+
+        // ** Rotate pid Sum to quad frame of reference, into pitch and roll **
+        float headingRads = (attitude.values.yaw / 10.0f) * RAD; // will be constrained to +/-pi in sin_approx()
+        const float sinHeading = sin_approx(headingRads);
+        const float cosHeading = cos_approx(headingRads);
+        float pidSumRoll = -sinHeading * nsPidSum + cosHeading * ewPidSum;
```

Comment:
(please try to follow conventions: https://en.wikipedia.org/wiki/Rotation_matrix)

Reply:
Very happy to change this to the appropriate convention, just explain what, exactly, I should change.

Obviously the signs of the distances and the PIDs and then these rotations all have to align.  Currently, they do.  So whatever change we make from where it is now must return the same sign at the output.  Hence, if we change these lines, we will have to either change the input direction or the signs of the PID elements.  

Other than that concern, I will change to whatever you suggest.

Reply:
Usual way to perform rotation is:
```   
    new_x = x * cos(phi) - y * sin(phi)
    new_y = x * sin(phi) + y * cos(phi)
```


For this to work, you need to convert heading to angle (phi is counterclockwise from X axis).

This was big problem in IMU - the equation was simply iterated until it worked (changing signs and swapping sin/cos), but it was almost impossible to change anything (and even iNAV did not fix it when refactoring everything else).